src/transport_ofi.c

```diff
@@ -1420,34 +1517,82 @@ int query_for_fabric(struct fabric_info *info)
+     * fabrics */
+    int num_nics = 0;
+    struct fi_info *fallback = NULL;
+    struct fi_info *filtered_fabrics_list_head = NULL;
+    struct fi_info *filtered_fabrics_last_added = NULL;
+    struct fi_info *multirail_fabric_list_head = NULL;
+    struct fi_info *multirail_fabric_last_added = NULL;
+
     if (info->fabric_name != NULL || info->domain_name != NULL) {
         struct fi_info *cur_fabric;
 
-        info->p_info = NULL;
-
         for (cur_fabric = info->fabrics; cur_fabric; cur_fabric = cur_fabric->next) {
             if (info->fabric_name == NULL ||
                 fnmatch(info->fabric_name, cur_fabric->fabric_attr->name, 0) == 0) {
                 if (info->domain_name == NULL ||
                     fnmatch(info->domain_name, cur_fabric->domain_attr->name, 0) == 0) {
-                    info->p_info = cur_fabric;
-                    break;
+                    if (!filtered_fabrics_list_head) filtered_fabrics_list_head = cur_fabric;
+                    if (filtered_fabrics_last_added) filtered_fabrics_last_added->next = cur_fabric;
+                    filtered_fabrics_last_added = cur_fabric;
                 }
             }
         }
+        filtered_fabrics_last_added->next = NULL;
     }
     else {
-        info->p_info = info->fabrics;
+        filtered_fabrics_list_head = info->fabrics;
+    }
+
+    struct fi_info *cur_fabric;
+
+    info->p_info = NULL;
+
+    if (shmem_internal_params.DISABLE_MULTIRAIL) {
+        info->p_info = filtered_fabrics_list_head;
     }
+    else {
+        /* Generate a linked list of all fabrics with a non-null nic value */
+        for (cur_fabric = filtered_fabrics_list_head; cur_fabric; cur_fabric = cur_fabric->next) {
+            if (!fallback) fallback = cur_fabric;
+            if (cur_fabric->nic && !nic_already_used(cur_fabric->nic, multirail_fabric_list_head, num_nics)) {
+                num_nics += 1;
+                if (!multirail_fabric_list_head) multirail_fabric_list_head = cur_fabric;
+                if (multirail_fabric_last_added) multirail_fabric_last_added->next = cur_fabric;
+                multirail_fabric_last_added = cur_fabric;
+            }
+        }
 
+        DEBUG_MSG("Total num. NICs detected: %d\n", num_nics);
+        if (num_nics == 0) {
+            info->p_info = fallback;
+        }
+        else {
+            int idx = 0;
+            struct fi_info **prov_list = (struct fi_info **) malloc(num_nics * sizeof(struct fi_info *));
+            for (cur_fabric = multirail_fabric_list_head; cur_fabric; cur_fabric = cur_fabric->next) {
+                prov_list[idx++] = cur_fabric;
+            }
+            qsort(prov_list, num_nics, sizeof(struct fi_info *), compare_nic_names);
+#ifdef USE_HWLOC
+            info->p_info = assign_nic_with_hwloc(info->p_info, prov_list, num_nics);
+#else
+            /* Round-robin assignment of NICs to PEs */
+            info->p_info = prov_list[shmem_internal_my_pe % num_nics];
```

Comment:
This might cause over-subscription in some cases, right?  Perhaps a FIXME or a github issue is needed to remind us that this would be better as `SHMEM_TEAM_NODE->my_pe % num_nics` if possible.  But it's too early in initialization to do that here...

Reply:
Added FIXME comment